src/nodes/hypertable_modify.c

```diff
@@ -745,7 +981,82 @@ ExecModifyTable(PlanState *pstate)
 	context.mtstate = node;
 	context.epqstate = &node->mt_epqstate;
 	context.estate = estate;
+	/*
+	 * For UPDATE/DELETE on compressed hypertable, decompress chunks and
+	 * move rows to uncompressed chunks.
+	 */
+	if ((operation == CMD_DELETE || operation == CMD_UPDATE) && !ht_state->comp_chunks_processed)
+	{
+		collect_chunks_from_scan(pstate, ht_state);
+		ListCell *ch;
+		ListCell *cf;
+		int pos = 0;
+		foreach (ch, ht_state->chunks)
+		{
+			bool chunk_status_changed = false;
+			PredicatesContext *cx = NULL;
+			List *scankeys = NIL;
+			List *null_columns = NIL;
+			List *is_null_checks = NIL;
 
+			if (ts_cm_functions->decompress_batches_for_update_delete)
+			{
+				if (ht_state->predicate_context)
+				{
+					cx = list_nth(ht_state->predicate_context, pos);
+					scankeys = cx->scankeys;
+					null_columns = cx->null_columns;
+					is_null_checks = cx->is_null_checks;
+					pos++;
+				}
+				if (ts_cm_functions->decompress_batches_for_update_delete(lfirst(ch),
+																		  scankeys,
+																		  null_columns,
+																		  is_null_checks,
+																		  &chunk_status_changed))
+				{
+					/*
+					 * tuples from compressed chunk has been decompressed and moved
+					 * to staging area, thus mark this chunk as partially compressed
+					 */
+					if (chunk_status_changed == true)
+						ts_chunk_set_partial(lfirst(ch));
+					/* free all chunks context */
+					if (scankeys)
+					{
+						foreach (cf, scankeys)
+							pfree((ScanKeyData *) lfirst(cf));
+					}
+					if (null_columns)
+					{
+						foreach (cf, null_columns)
+							pfree((Bitmapset *) linitial(null_columns));
```

Comment:
```suggestion
							pfree(lfirst(cf));
```

Probably should be like this? Also no need to cast because pfree takes `void *`. The list probably also should be emptied so that it doesn't contain the invalid elements. Maybe this can be replaced entirely by `list_free_deep`.

Reply:
This code is no longer valid.